### PR TITLE
Parse file included inside a macro at definition time

### PR DESCRIPTION
**Pull request by ReallyNiceGuy**
Fixes issue number #1473.
The contents of the included file will be parsed at macro definition time, as if it is part of the current file.

### CODE REVIEW

**Comment by spiro-trikaliotis** [comment]
Are you really sure that your patch works without breaking anything else (other than making #231 fail to compile)?

I am currently working on the parser, too, in order to fix #1427. What I found out is that the behaviour of the parser is rather complicated, and it is rather hard to follow.

To me, your solution looks "too easy". Furthermore, you add include file processing at a second point in the sources, that is, you add some type of redundancy. To me, it does not look right, although I cannot tell exactly why at the moment.

**Comment by ReallyNiceGuy** [comment]
> Are you really sure that your patch works without breaking anything else (other than making #231 fail to compile)?
> 
At least the automated tests pass. As it wasn't working at all before, I have a strong belief that it won't break any sensible code.

> I am currently working on the parser, too, in order to fix #1427. What I found out is that the behaviour of the parser is rather complicated, and it is rather hard to follow.
I completely agree with you.

> 
> To me, your solution looks "too easy".
At this point, it is very easy, because I just start injecting the contents of the included file on the macro itself.

> Furthermore, you add include file processing at a second point in the sources, that is, you add some type of redundancy. To me, it does not look right, although I cannot tell exactly why at the moment.
I thought about it and could not find a sane way to follow the normal path. The DoInclude function is working in normal token mode, but in the macro processing, it is in RawTokenMode.

**Comment by oliverschmidt** [comment]
> At least the automated tests pass. [...]

The tests focus almost exclusively on the C compiler so that doesn't say much :-(

**Comment by oliverschmidt** [comment]
Any additional thoughts on how to move forward with this PR?

**Comment by ReallyNiceGuy** [comment]
> Any additional thoughts on how to move forward with this PR?

I was thinking about my own comment and it doesn't make much sense to have the include work at expansion time.
Otherwise, the parsing would have to be done incrementally, and I doubt that this is what was intended.

**Comment by oliverschmidt** [comment]
> [...] doesn't make much sense to have the include work at expansion time.

I can't really follow you :-( Please elaborate.

**Comment by ReallyNiceGuy** [comment]
I said above:

>  It could be improved by including the file at the point of use, but, to be honest, I didn't figure out on how to implement it.

But then I thought about it and concluded that this idea is too complicated.
The question we have to ask is: what does it mean to have an .include inside a macro.
The implementation in this patch is:
  pull the contents of the include and put inside the macro, as if it was in the same source file. What follows is the that rules apply as if you had typed the file in at that point (a new macro definition will not work, for example)

The option is:
  pull the contents of the include file only at the expansion of the macro. This would allow the issue #231 to work as the reporter intended, but now you can define a macro inside the expansion of a macro, inside the expansion of a macro, etc. I am not sure if this is what .include was intended for.

I hope this makes sense.

**Comment by spiro-trikaliotis** [comment]
The behaviour of ca65 with macros is as follows:
When parsing a macro, it convert them to tokens and stores them. Other than that, it does nothing. Most importantly, it does not interpret any token other than the corresponding .endmacro, or an .local (for traditional/classic macros only).

That's all done inside of the while (1) loo where @ReallyNiceGuy added his .include processing.

That's why I have my doubts with this solution:

1. Does it break non-classic macros?
2. As everything is processed at use time, why not this? What are the side effects?

I did some test and immediately found a regression. Have a look at the attached zip and the files a1.a65 and a2.65.

In a1.a65, there is a .repeat inside of the .macro. It works flawlessly.
In a2.a65, the .repeat inside of the .macro has been put into a.inc instead. Note that a1.a65 and a2.a65 result in the identical commands once tokenized (ommitting the .include directive itself).

While a2.a65 assembles witthout this PR, it does not with this PR. The .repeat instructions are not recognized. They are not recognized because they are handled now at definition time, and the loop does not know about the .repeat. They must be handled at use time, as they are in the original code.

[broken.tar.gz](https://github.com/cc65/cc65/files/6406699/broken.tar.gz)

**Comment by ReallyNiceGuy** [comment]
This is the compilation with the changes in this PR with your slightly modified examples.
Notice that the compilation result is exactly the same.
[pr_test.tar.gz](https://github.com/cc65/cc65/files/6406857/pr_test.tar.gz)

**Comment by ReallyNiceGuy** [comment]
> @ReallyNiceGuy: You must slow down a little bit and give me time to react... This is no chat here. ;)

Sorry, I just posted as soon as I got a test finished. I don't expect an immediate reaction, no worries there.

> You are right, I managed to confuse the "original" (cc65's git master) ca65 and your version. So, to correct my statement: Your version assembles the file, the original does not. So, your code is an improvement.
> 
> I am not sure how and why this happened. I must have had a bad brain at that moment.

It happens to everybody, also, no worries.
> 
> So, it seems your version is better than the original. I will do some additional tests to find out what happens if you nest macros and includes (and macros in the includes and includes in that macro and so on).

Nested macros defined inside the include will fail to compile, in the same manner as nested macros in the same source file.

> 
> I still believe handling it at usage time would be the better approach, as we are splitting the logic between two places now, which will make the code less manageable, IMHO. It is already very complicated.

I only handle the TOK_INCLUDE and the subsequent string with the file name, nothing else.
The issue I was having with the expansion code was that it was way more intrusive. I would have to start getting tokens from a file while in the middle of the expansion, and it was not pretty. Handling at the definition time was easy, and the only things I can think of that we miss with this approach are:
1. I cannot have macro definitions in the files included inside a macro (fine by me, might cause havoc if allowed)
2. I cannot use a macro to include a file defined at expansion time by an argument to that macro (maybe there are use cases for this one, issue #231 comes to mind), but I don't see it as that bad.

**Comment by spiro-trikaliotis** [comment]
You are right, I overlooked that nested macros are not possible now, too. So, this is not a regression.

I tested some more things. I did not find any problems with this patch. I still don't like the approach of handling it at definition time, but I also see your point that it would be harder to do it right at expansion time.

From my point of view (if this matters at all!), this patch can be merged.

**Comment by oliverschmidt** [comment]
> From my point of view (if this matters at all!), this patch can be merged.

@spiro-trikaliotis: Thank you very much for the thoughts you put into assessing this PR! In fact it matters a lot as it triggers me merging it ;-)

@ReallyNiceGuy: Thanks for being open to answer all questions coming up.

**Comment by oliverschmidt** [comment]
I'm wondering if https://sourceforge.net/p/cc65/mailman/message/37306831/ is a result of merging this PR. The message says:

```
wanted to do some work with the ancheron VM...started with the basics...
makefile is given which uses ca65
but I am on windows.  
gnu make installed.
cc65 installed.  

I run make which would execute:
rm -r obj
rm -r bin
mkdir  obj
mkdir  bin
ca65 -l obj/acheron.o.lst src/acheron.asm -g -o obj/acheron.o -I src
ld65 -Ln bin/labels -C src/acheron.cfg -m obj/map -o bin/acheron.prg obj/acheron.o
# Display size information for each component. Only the above line is necessary for the build.
grep '^[^_]* Size=' obj/map
grep '\._size__' bin/labels
dir bin/acheron.prg

but at the ca65 step I get this:
rm -r obj
RemoveDirectory(C:\retro\6502\acheronvm-master\acheronvm-master\obj)
rm -r bin
RemoveDirectory(C:\retro\6502\acheronvm-master\acheronvm-master\bin)
mkdir  obj
mkdir  bin
ca65 -l obj/acheron.o.lst src/acheron.asm -g -o obj/acheron.o -I src
src/acheron.asm:38: Error: String constant expected
115 / 128 opcodes defined
make: *** [obj/acheron.o] Error 1

it's having a issue with this macro:
.macro reportSize label, filename
 label:
 .include filename
 .ident (.concat("_size__", .string(label))) = * - name
 .export .ident (.concat("_size__", .string(label)))
 .endmacro


specifically the .include filename

Any ideas?  
```

@ReallyNiceGuy: What do you think?

**Comment by ReallyNiceGuy** [comment]
Yes, it is. The error is caused by the .include being resolved at definition time, not at usage time.

I will check what was the result before the patch and how to modify the macro to make it work as expected again.

**Comment by ReallyNiceGuy** [comment]
He can remove this macro completely, it is not used anywhere.

**Comment by ReallyNiceGuy** [comment]
I created a PR to the project directly.

**Comment by oliverschmidt** [comment]
> I created a PR to the project directly.

I really appreciate your super-quick and very helpful reaction! Thanks :-)

**Comment by ReallyNiceGuy** [comment]
👍
